proposals/excess_reserve_capacity.md

```diff
@@ -0,0 +1,49 @@
+
+* Only Job of MachineTaint controller will be:
+    * List all the Machines under each MachineDeployment.
+    * Put taints of [*noSchedule* and *noExecute*](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) on latest MachineObjects.
+        * There should be an event-based informer mechanism where MachineTaintController gets to know about any Update/Delete/Create event of MachineObjects - in turn, maintains the *noSchedule* and *noExecute* taints on all the *latest* machines.
+        * Why latest machines?
+            * Whenever autoscaler decides to add new machines - essentially ScaleUp event - taints from the older machines are removed and newer machines get the taints. This way X number of Machines immediately becomes free for new pods to be scheduled.
+            * While ScaleDown event, autoscaler specifically mentions which machines should be deleted, and that should not bring any concerns. Though we will have to put proper label/annotation defined by autoscaler on taintedMachines, so that autoscaler does not consider the taintedMachines for deletion while scale-down.
+                * Annotation on tainted node: `"cluster-autoscaler.kubernetes.io/scale-down-disabled": "true"`
+* Implementation Details:
+    * Expect new **optional field** *ExcessReplicas* in `MachineDeployment.Spec`. MachineDeployment controller now adds both `Spec.Replicas` and `Spec.ExcessReplicas`[if provided], and considers that as a standard desiredReplicas.
+        * Current working of MCM will not be affected if ExcessReplicas field is kept nil.
+    * MachineController currently reads the *NodeObject* and sets the MachineConditions in MachineObject. Machine-controller will now also read the taints/labels from the MachineObject - and maintains it on the *NodeObject*.
+
+* We expect cluster-autoscaler to intelligently make use of the provided feature from MCM.
+    * CA gets the input of *min:max:excess* from Gardener. CA continues to set the `MachineDeployment.Spec.Replicas` as usual based on the application-workload.
+    * In addition, CA also sets the `MachieDeployment.Spec.ExcessReplicas` .
+    * Corner-case:
+        * CA should decrement the excessReplicas field accordingly when *desiredReplicas+excessReplicas* on MachineDeployment goes beyond *max*.
+
+
+### Approach 2: Enhance Cluster-autoscaler by simulating fake pods in it.
+* There was already an attempt by community to support this feature.
+    * Refer for details: https://github.com/kubernetes/autoscaler/pull/77/files
+
+### Approach 3: Enhance cluster-autoscaler to support pluggable scaling-algorithms.
+* Forked version of cluster-autoscaler could be improved to plug-in the algorithm for excess-reserve capacity.
+* Needs further discussion around upstream support.
+
+### Approach 4: Make intelligent use of Low-priority pods.
```

Comment:
Approach 4 is also a hack like approach 2. But at least it has a chance to work because it is not dependent on the closed PR.

But this would be workable only as a short-term solution only for seed clusters. For shoot clusters we need a more general solution.

Reply:
One minor disadvantage of this approach would also be that it will put a restriction on users to schedule their own low(-est) priority pods because they will compete with these pods.